### PR TITLE
Full description of Application using `describe` subcommand

### LINKED ISSUE #197: get a full description of my application
As a developer, I want to be able to get a full description of my application

```
$ ocd app describe
Application XYZ has:
Component ABC with source in https://git and scaled to 2 instances, exposed via http://blabla and connected to component DB1, and component DB2
Component DB1, based of PostgreSQL and scaled to 1 connected by Component ABC
Component DB2, based of MongoDB and scaled to 1 connected by Component ABC
```

Full description would entail:
- name of the application 
- description of each component the application has.
  - Each component's description should provide:
    - name
    - source location
    - number of instances
    - connections (urls, service used)

Subtask of #155

### PR DESCRIPTION
**Pull request by surajnarwade**
Resolves #197

It will give output as following:

```
ocdev application describe batman
Application batman  has:
Component nodejs with source in file:///tmp/nodejs-ex
This Component is externally exposed via nodejs-myproject.192.168.42.231.nip.io
This component uses storage which is mounted at /home
```

### CODE REVIEW

**Comment by surajnarwade** [comment]
@kadel @containscafeine , feedback needed here :)

File: cmd/application.go [hunk 1]
```diff
@@ -167,6 +172,38 @@ var applicationSetCmd = &cobra.Command{
...
 }
 
+var applicationDescribeCmd = &cobra.Command{
+	Use:   "describe",
+	Short: "describe the given application",
+	Run: func(cmd *cobra.Command, args []string) {
+		client := getOcClient()
+		fmt.Println("Application", args[0], " has:")
+
+		a, _ := component.List(client)
```
Comment (kadel) [inline cmd/application.go]: you need to check for errors, here and in other places.

But otherwise, this looks good.

Comment (surajnarwade) [reply cmd/application.go]: yeah

**Comment by surajnarwade** [comment]
thanks @slemeur @kadel for the feedback :)

File: cmd/application.go [hunk 2]
```diff
@@ -167,6 +172,38 @@ var applicationSetCmd = &cobra.Command{
...
+		c, _ := project.GetCurrent(client)
```
Comment (kadel) [inline cmd/application.go]: also please don't use one letter variable names

File: cmd/application.go [hunk 3]
```diff
@@ -167,6 +172,54 @@ var applicationSetCmd = &cobra.Command{
...
 
+var applicationDescribeCmd = &cobra.Command{
+	Use:   "describe",
+	Short: "describe the given application",
+	Run: func(cmd *cobra.Command, args []string) {
+		client := getOcClient()
+		fmt.Println("Application", args[0], " has:")
+
+		componentList, err := component.List(client)
+		if err != nil {
```
Comment (containscafeine) [inline cmd/application.go]: replace with `checkError()`

File: cmd/application.go [hunk 4]
```diff
@@ -167,6 +172,54 @@ var applicationSetCmd = &cobra.Command{
...
+			fmt.Println(err)
+			os.Exit(1)
+		}
+		currentProject, err := project.GetCurrent(client)
+		if err != nil {
```
Comment (containscafeine) [inline cmd/application.go]: replace with `checkError()`

File: cmd/application.go (hunk 3 above)
Comment (containscafeine) [reply cmd/application.go]: #277

File: cmd/application.go [hunk 5]
```diff
@@ -167,6 +172,54 @@ var applicationSetCmd = &cobra.Command{
 	},
 }
 
+var applicationDescribeCmd = &cobra.Command{
+	Use:   "describe",
+	Short: "describe the given application",
+	Run: func(cmd *cobra.Command, args []string) {
+		client := getOcClient()
+		fmt.Println("Application", args[0], " has:")
```
Comment (containscafeine) [inline cmd/application.go]: I think `fmt.Println()` enters a whitespace anyway, so we don't need it before `has:`
Also, it'd be better to format this with `fmt.Printf()`, looks cleaner and like an entire sentence.

File: cmd/application.go [hunk 6]
```diff
@@ -139,6 +144,50 @@ var applicationSetCmd = &cobra.Command{
...
+
+		componentList, err := component.List(client)
+		if err != nil {
+			fmt.Println(err)
+			os.Exit(1)
+		}
+		currentProject := project.GetCurrent(client)
```
Comment (kadel) [inline cmd/application.go]: Would it make sense to move this to  `application.Describe`. so we use that function in other places like displaying info before the application gets deleted?

The downside might be that `application.Describe` will have to print to stdout.
WDYT @containscafeine?

**Comment by kadel** [comment]
currently this is blocked on #295

**Comment by containscafeine** [comment]
blocked on #289

File: cmd/application.go [hunk 7]
```diff
@@ -139,6 +141,55 @@ var applicationSetCmd = &cobra.Command{
 	},
 }
 
+var applicationDescribeCmd = &cobra.Command{
+	Use:   "describe",
```
Comment (containscafeine) [inline cmd/application.go]: add `[component name]`

File: cmd/application.go [hunk 8]
```diff
@@ -139,6 +141,55 @@ var applicationSetCmd = &cobra.Command{
...
+	Short: "describe the given application",
+	Example: `  # Describe someapp application,
```
Comment (containscafeine) [inline cmd/application.go]: replace someapp with a different app name, webapp, or nodejsapp

File: cmd/application.go [hunk 9]
```diff
@@ -139,6 +141,55 @@ var applicationSetCmd = &cobra.Command{
 	},
 }
 
+var applicationDescribeCmd = &cobra.Command{
```
Comment (containscafeine) [inline cmd/application.go]: add `args: maxArgs(1)`

File: cmd/application.go [hunk 10]
```diff
@@ -139,6 +141,55 @@ var applicationSetCmd = &cobra.Command{
...
+			var err error
+			currentApplication, err = application.GetCurrent(client)
+			checkError(err, "")
+		} else {
+			currentApplication = args[0]
+			//Check whether application exist or not
+			exists, err := application.Exists(client, currentApplication)
+			checkError(err, "")
+			if !exists {
+				fmt.Printf("application with the name %s does not exist\n", currentApplication)
```
Comment (containscafeine) [inline cmd/application.go]: Capitalize first character

File: cmd/application.go [hunk 11]
```diff
@@ -139,6 +141,55 @@ var applicationSetCmd = &cobra.Command{
...
+				os.Exit(1)
+			}
+		}
+		fmt.Printf("Application %s has:", currentApplication)
```
Comment (containscafeine) [inline cmd/application.go]: add new line after `has:`

File: cmd/application.go [hunk 12]
```diff
@@ -139,6 +141,55 @@ var applicationSetCmd = &cobra.Command{
...
+		//Project
+		currentProject := project.GetCurrent(client)
+		// List of Component
+		componentList, err := component.List(client, currentApplication, currentProject)
```
Comment (containscafeine) [inline cmd/application.go]: If `len(componentList) == 0`, print a message saying that "This application has no components deployed."

File: cmd/application.go [hunk 13]
```diff
@@ -139,6 +141,55 @@ var applicationSetCmd = &cobra.Command{
...
+			}
+			// URL
+			if componentURL != "" {
+				fmt.Println("This Component is externally exposed via", componentURL)
+			}
+			// Storage
+			for _, store := range appStore {
+				fmt.Println("This Component uses storage", store.Name, "of size", store.Size)
+			}
+
```
Comment (containscafeine) [inline cmd/application.go]: remove extra newline

File: cmd/application.go (hunk 9 above)
Comment (surajnarwade) [reply cmd/application.go]: added

File: cmd/application.go [hunk 14]
```diff
@@ -139,6 +141,60 @@ var applicationSetCmd = &cobra.Command{
...
+		componentList, err := component.List(client, currentApplication, currentProject)
+		checkError(err, "")
+		if len(componentList) == 0 {
+			fmt.Printf("Application %s has no components deployed.\n", currentApplication)
+			os.Exit(1)
+		}
+		fmt.Printf("Application %s has:\n", currentApplication)
+
+		for _, cmpnt := range componentList {
+			componentType, path, componentURL, appStore, err := component.GetComponentDesc(client, cmpnt.Name, currentApplication, currentProject)
```
[the remark this answers is not in the document]
Comment (surajnarwade) [reply cmd/application.go]: done

File: cmd/util.go [hunk 15]
```diff
@@ -24,3 +26,18 @@ func getComponent(client *occlient.Client, inputComponent, applicationName, proj
 	}
 	return inputComponent
 }
+
+func PrintComponentInfo(cmpntName string, componentType string, path string, componentURL string, appStore []storage.StorageInfo) {
```
Comment (containscafeine) [inline cmd/util.go]: we don't need to export this

Comment (containscafeine) [inline cmd/util.go]: add comment here

File: cmd/application.go [hunk 16]
```diff
@@ -139,6 +141,50 @@ var applicationSetCmd = &cobra.Command{
...
+			checkError(err, "")
+			PrintComponentInfo(cmpnt.Name, componentType, path, componentURL, appStore)
+
```
Comment (containscafeine) [inline cmd/application.go]: nit: extra newline here

**Pull request merged**